wiki/python/ESTIMATOR_GUIDE.md

```diff
@@ -0,0 +1,866 @@
+   ```python
+   from cuml.common.array_descriptor import CumlArrayDescriptor
+
+   class MyEstimator(Base):
+
+      labels_ = CumlArrayDescriptor()
+
+      def __init__(self):
+         ...
+   ```
+5. Add input and return type annotations to public API functions OR wrap those functions explicitly with conversion decorators (see [Estimator Methods](#estimator-methods) for non-standard use cases)
+   ```python
+   class MyEstimator(Base):
+      
+      def fit(self, X) -> "MyEstimator":
+         ...
+
+      def predict(self, X) -> CumlArray:
+         ...
+   ```
+6. Implement `get_param_names()` including values returned by `super().get_param_names()`
+   ```python
+   def get_param_names(self):
+      return super().get_param_names() + [
+         "eps",
+         "min_samples",
+      ]
+   ```
+
+For the majority of estimators, the above steps will be sufficient to correctly work with the cuML library and ensure a consistent API. However, situations may arrise where an estimator differs from the standard pattern and some of the functionality needs to be customized. The remainder of this guide takes a deep dive into the estimator functionality to assist developers when building estimators.
+
+## Background
+
+Some background is necessary to understand the design of estimators and how to work around any non-standard situations.
+
+### Input and Output Types in cuML
+
+In cuML we support both ingesting and generating a variety of different n-dimensional array types. Estimators should be able to accept and return any array type. The types that are supported as of release 0.17:
+
+ - cuDF DataFrame or Series
+ - Pandas DataFrame or Series
+ - NumPy Arrays
+ - Numba Device Arrays
+ - CuPy arrays
+ - CumlArray type (Internal to the `cuml` API only.)
+
+When converting between types, it's important to minimize the CPU<->GPU type conversions as much as possible. Conversions such as NumPy -> CuPy or Numba -> Pandas DataFrame will incur a performance penalty as memory is copied from device to host or viceversa.
+
+Converting between types of the same device, i.e. CPU<->CPU or GPU<->, do not have as significant of a penalty, though they may still increase memory usage.
+
+Finally, conversions between Numba<->CuPy<->CumlArray incur the least amount of overhead since only the device pointer is moved from one class to another.
+
+Internally, all arrays should be converted to `CumlArray` as much as possible since it is compatible with all output types and can be easily converted.
+
+### Speficying the Array Output Type
+
+Users can choose which array type should be returned by cuml by either: 
+1. Individually setting the output_type property on an estimator class (i.e `Base(output_type="numpy")`)
+2. Globally setting the `cuml.global_output_type`
+3. Temporarily setting the `cuml.global_output_type` via the `cuml.using_output_type` context manager
```

Comment:
I don't think it is needed to state here that the context manager sets the global output_type, that's an internal implementation detail. The context manager works like a typical python context manager.

Reply:
Are you suggesting I should remove `#3`?

Reply:
Ok, resolving.